File: code/ex702.java
class Solution {
    public int search(ArrayReader reader, int target) {
        int max = 2147483647;
        if (reader.get(0) == max)    return -1;
        
        // Find the end point of the array
        int lo = 0, hi = Integer.MAX_VALUE;
        int end = 0;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (reader.get(mid) == max) {
                if (reader.get(mid - 1) != max) {
                    end = mid - 1;
                    break;
                } else {
                    hi = mid - 1;
                }
            } else {
                lo = mid + 1;
            }
        }
        
        // Find the target
        lo = 0; hi = end;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (reader.get(mid) == target)    return mid;
            else if (reader.get(mid) > target) {
                hi = mid - 1;
            }
            else {
                lo = mid + 1;
            }
        }
        
        return -1;
    }
}

File: code/ex702.cpp
/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        int lo = 0, hi = 1e4;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int val = reader.get(mid);
            if (val == target)    return mid;
            else if (val > target)    hi = mid - 1;
            else    lo = mid + 1;
        }
        return -1;
    }
};